### PR TITLE
feat!: add `from_dict` method to construct configuration from dictionary

### LINKED ISSUE #14: [Enhancement]: Add support for creating Slurm configuration from dictionary
By adding a `from_dict` method, it makes it much easier to work with Slurm configuration information as configuration objects do not need to be constructed manually. The `from_dict()` constructor can take in data passed in from a Juju integration in bulk and make quick changes to the relevant configuration file.

The alternative constructor should function like the following:

```python3
from slurmutils.models import Node

node = Node.from_dict(
    {
        "NodeName": "batch-[0-25]", 
        "NodeAddr": "12.34.56.78", 
        "CPUs": 1, 
        "RealMemory": 1000, 
        "TmpDisk": 10000,
    }
)
print(node.node_name)
```

### PR DESCRIPTION
**Pull request by NucciTheBoss**
TL;DR - adds the `from_dict` class method to all the configuration data models to enable creating new model objects from dictionaries. For example, here's how you could create a node using the `from_dict` method:

```python3
new_node = Node.from_dict(
    {
        "juju-c9fc6f-5": {
            "NodeAddr": "10.152.28.51",
            "CPUs": "1",
            "RealMemory": "1000",
            "TmpDisk": "10000",
        }
    }
)
```

The new `option` module is then responsible for validating that the values passed to `from_dict` are correct, add provides the marshallers for when you go to dump the configuration into a file. The `option` module also removes the "hacky" implementation for callbacks which just uses a `MappingProxy` embedded into the data model as a protected attribute.

BREAKING CHANGES: I removed the `NodeMap`, `PartitionMap`, `DownNodesList` wrappers as I was not enamored with their implementation. _They were rushed_. I took them out as I'd like to spend more time refining their implementation so they aren't as hacky/clunky. Let me know if you'd like for me to add them back in. We can like open a follow-on PR that focuses on their implementation, but for now we have enough to unblock https://github.com/charmed-hpc/hpc-libs/pull/22

> Side note: I need to update the README and pyproject.toml before merging this PR, but I want to know your thoughts on the "`from_dict(...)` and back again" implementation before going ahead and handling that.

Closes #14

### CODE REVIEW

**Review by jedel1043 — APPROVED** [review]
Looking good! Just found some typos but implementation-wise looks perfect!

File: slurmutils/editors/editor.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    """Clean line before further processing.
+
+    Returns:
+        Returns the cleaned line and False if it should be ignored.
+        If True, then the processors should ignore the line.
+    """
+    if _is_comment(line):
+        return "", True
+
+    return (_slice_comment(line) if _contains_comment(line) else line).strip(), False
```
Comment (jedel1043) [inline slurmutils/editors/editor.py:53]: Now that I read this again, does this need all the guards? I'm seeing that `_slice_comment` will already return the correct result in all cases, even if the line starts with "#" or if the line has a trailing comment.

File: slurmutils/editors/editor.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    """Slice inline comment off of line."""
+    return line.split("#", maxsplit=1)[0]
+
+
+def clean(line: str) -> Tuple[str, bool]:
+    """Clean line before further processing.
+
+    Returns:
+        Returns the cleaned line and False if it should be ignored.
+        If True, then the processors should ignore the line.
```
Comment (jedel1043) [inline slurmutils/editors/editor.py:48]: This distinguishes between lines that start with "#" and lines that start with whitespace, but also have a comment at the end (" # indented comment"), making it throw an error in the second case (probably). Maybe we should just use the empty string as our marker to indicate that the line should be ignored.

File: slurmutils/models/model.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    """
+
+    def getter(self):
+        return self.data.get(opt, None)
+
+    def setter(self, value):
+        self.data[opt] = value
+
+    def deleter(self):
+        del self.data[opt]
```
Comment (jedel1043) [inline slurmutils/models/model.py:66]: Praise: nice usage of closures!

File: slurmutils/models/model.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+       For example, without additional preprocessing, the key `CPUs` will
+       become `cp-us` if put through a caramelize with being preformatted to `Cpus`.
+    """
+    if "CPUs" in key:
+        key = key.replace("CPUs", "Cpus")
+    key = _acronym.sub(r"_", key)
+    return _camelize.sub(r"_", key).lower()
+
+
+def generate_descriptors(opt: str):
```
Comment (jedel1043) [inline slurmutils/models/model.py:52]: Type annotation?

File: slurmutils/models/model.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]):
+        """Construct new model from dictionary."""
+        return cls(**data)
+
+    @classmethod
+    def from_json(cls, obj: str):
+        """Construct new model from JSON object."""
+        data = json.loads(obj)
+        return cls.from_dict(data)
```
Comment (jedel1043) [inline slurmutils/models/model.py:108]: Nit: this can be typed with a typevar:
```Python
TBaseModel = TypeVar("TBaseModel", bound="BaseModel")

def from_dict(cls: Type[TBaseModel], data: Dict[str, Any]) -> TBaseModel
```
But we can also skip this if you think it's getting too weird for just type annotations 😅

File: slurmutils/models/option.py [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+    Callback,
+    ColonSeparatorCallback,
+    CommaSeparatorCallback,
+    ReasonCallback,
+    SlurmDictCallback,
+)
+
+
+@dataclass(frozen=True)
+class _Option:
```
Comment (jedel1043) [inline slurmutils/models/option.py:40]: Nit: seems weird to call this `_Option` when it's the base class for a list of options, so maybe something like `_OptionsList` could be better?

**Review by jedel1043 — COMMENTED** [review]
Did another pass with some questions and some nitpicks.

File: slurmutils/models/option.py (hunk 6 above)
Comment (NucciTheBoss) [reply slurmutils/models/option.py:40]: How about `_OptionSet`?

I originally had the name to be `_OptionList`, but I changed it to just `_Option` because "List" implies to me that it's an ordered data structure when really it's just a set of valid options for various configuration objects with callbacks for certain configuration option values.

Comment (jedel1043) [reply slurmutils/models/option.py:40]: Yeah, `_OptionSet` sounds good!

File: slurmutils/models/model.py (hunk 5 above)
Comment (NucciTheBoss) [reply slurmutils/models/model.py:108]: I'd consider this as out of scope for this PR. 

`slurmutils` overall needs proper type checking added to it, so I'm expecting that this won't be the only place where we need to add better and more accurate type annotations :sweat_smile:

File: slurmutils/models/model.py (hunk 4 above)
Comment (NucciTheBoss) [reply slurmutils/models/model.py:52]: Yeah - I can add one, just returns tuple of callables.

File: slurmutils/editors/editor.py (hunk 2 above)
Comment (NucciTheBoss) [reply slurmutils/editors/editor.py:48]: I prefer to have the boolean flag to indicate whether the line should be processed or not. I wanted a similar implementation like Go maps where:

```go
v, ok = m["NodeName"]
if !ok {
        fmt.Println("Oops - forgot the node name")
       os.Exit(1)
}
```

That way we don't need to rely on a default value for the primary argument, and we can just choose to ignore the second return if we know that the output we want is good.

File: slurmutils/editors/editor.py (hunk 1 above)
Comment (NucciTheBoss) [reply slurmutils/editors/editor.py:53]: Hmm... I originally used `_is_comment(...)` to quickly check if a line is comment and we shouldn't worry about even performing the slice. I can reduce this implementation to see if `_is_comment(...)` is really necessary.

File: slurmutils/editors/editor.py (hunk 2 above)
Comment (jedel1043) [reply slurmutils/editors/editor.py:48]: Then `Optional[str]` should be enough in that case, returning `None` if the line shouldn't be processed. I don't think we need to rely on go-isms since Python natively supports optionals.

Comment (NucciTheBoss) [reply slurmutils/editors/editor.py:48]: _But but... MY GOISMS!! :exploding_head:_

Good compromise for me :+1:

File: slurmutils/editors/editor.py (hunk 1 above)
Comment (NucciTheBoss) [reply slurmutils/editors/editor.py]: Removed `_is_comment(...)` in https://github.com/charmed-hpc/slurmutils/pull/15/commits/87e6fb2c690a86e6fccdd662acdf5806da72648c

File: slurmutils/editors/editor.py (hunk 2 above)
Comment (NucciTheBoss) [reply slurmutils/editors/editor.py]: Less Go-ish in https://github.com/charmed-hpc/slurmutils/pull/15/commits/87e6fb2c690a86e6fccdd662acdf5806da72648c

File: slurmutils/models/option.py (hunk 6 above)
Comment (NucciTheBoss) [reply slurmutils/models/option.py]: Implemented `_OptionSet` and better naming overall for the configuration option sets in https://github.com/charmed-hpc/slurmutils/pull/15/commits/703303ff30e6e3631c9c0320376670d33ddcf257

File: slurmutils/models/model.py (hunk 4 above)
Comment (NucciTheBoss) [reply slurmutils/models/model.py]: Added type annotation in https://github.com/charmed-hpc/slurmutils/pull/15/commits/67a1bdc9be8fb1b5ea42cba586883cdb6465a183. Further work will need to be done in a separate PR to enable project wide type checking. I did initially set up pyright, but it complained quite a bit and the dynamically defined descriptors on the data models :sweat_smile:

**Comment by NucciTheBoss** [comment]
@jedel1043 R4R 🔍

**Review by jedel1043 — APPROVED** [review]
Looks great!

**Comment by jamesbeedy** [comment]
Wow! @NucciTheBoss , nice work!
